drivers/dma/dma_andes_atcdmac300.c

```diff
@@ -0,0 +1,428 @@
+				break;
+			case DMA_ADDR_ADJ_NO_CHANGE:
+				ch_ctrl |= DMA_CH_CTRL_SRCADDR_FIX;
+				break;
+			default:
+				ret = -EINVAL;
+				goto end;
+			}
+
+			switch (cfg_blocks->dest_addr_adj) {
+			case DMA_ADDR_ADJ_INCREMENT:
+				ch_ctrl |= DMA_CH_CTRL_DSTADDR_INC;
+				break;
+			case DMA_ADDR_ADJ_DECREMENT:
+				ch_ctrl |= DMA_CH_CTRL_DSTADDR_DEC;
+				break;
+			case DMA_ADDR_ADJ_NO_CHANGE:
+				ch_ctrl |= DMA_CH_CTRL_DSTADDR_FIX;
+				break;
+			default:
+				ret = -EINVAL;
+				goto end;
+			}
+			dma_chain[channel][current_block_idx].ctrl = ch_ctrl;
+			dma_chain[channel][current_block_idx].transize =
+						cfg_blocks->block_size/cfg->source_data_size;
+
+			dma_chain[channel][current_block_idx].srcaddrl =
+						(uint32_t)cfg_blocks->source_address;
+			dma_chain[channel][current_block_idx].srcaddrh = 0x0;
+
+			dma_chain[channel][current_block_idx].dstaddrl =
+						(uint32_t)((long)cfg_blocks->dest_address);
+			dma_chain[channel][current_block_idx].dstaddrh = 0x0;
+
+			if (cfg_blocks->next_block) {
+				dma_chain[channel][current_block_idx].llpointerl =
+					(uint32_t)&dma_chain[channel][current_block_idx + 1];
+				dma_chain[channel][current_block_idx].llpointerh = 0x0;
+
+				current_block_idx = current_block_idx + 1;
+
+			} else {
+				dma_chain[channel][current_block_idx].llpointerl = 0x0;
+				dma_chain[channel][current_block_idx].llpointerh = 0x0;
+				dma_chain[channel][current_block_idx].next_block = NULL;
+			}
+		}
+	} else {
+		/* Single transfer is supported, but Chain transfer is still
+		 * not supported. Therefore, set LLPointer to zero
+		 */
+		sys_write32(0, DMA_CH_LL_PTR_L(dev, channel));
+		sys_write32(0, DMA_CH_LL_PTR_H(dev, channel));
+	}
+
+end:
+	k_spin_unlock(&data->chan[channel].lock, key);
+	return ret;
+}
+
+static int dma_atcdmac300_reload(const struct device *dev, uint32_t channel,
+				 uint32_t src, uint32_t dst, size_t size)
+{
+	struct dma_atcdmac300_data *const data = dev->data;
+	uint32_t src_width;
+	k_spinlock_key_t key;
+
+	if (channel >= ATCDMAC100_MAX_CHAN) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&data->chan[channel].lock);
+
+	/* Set source and destination address */
+	sys_write32(src, DMA_CH_SRC_ADDR_L(dev, channel));
+	sys_write32(0, DMA_CH_SRC_ADDR_H(dev, channel));
+	sys_write32(dst, DMA_CH_DST_ADDR_L(dev, channel));
+	sys_write32(0, DMA_CH_DST_ADDR_H(dev, channel));
+
+	src_width = sys_read32(DMA_CH_CTRL(dev, channel)) & DMA_CH_CTRL_SWIDTH_MASK;
+	src_width = (BIT(src_width >> DMA_CH_CTRL_SWIDTH_POS));
+
+	/* Set transfer size */
+	sys_write32(size/src_width, DMA_CH_TRANSIZE(dev, channel));
+
+	k_spin_unlock(&data->chan[channel].lock, key);
+
+	return 0;
+}
+
+static int dma_atcdmac300_transfer_start(const struct device *dev,
+					 uint32_t channel)
+{
+	struct dma_atcdmac300_data *const data = dev->data;
+	k_spinlock_key_t key;
+
+	if (channel >= ATCDMAC100_MAX_CHAN) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&data->chan[channel].lock);
```

Comment:
Is there a particular reason to have a per channel spin lock? The semantics of DMA in Zephyr are predominantly single owner, as in a single owner is responsible for each channel.

If you need to have share a channel among call contexts its up to the caller to provide synchronization.

Reply:
Oh, I got it.
We have removed the spin_lock for per channel.
Use the common spin_lock of dma driver to handle the critical section of dma common register of all channels.

Thanks a lot for this request.

Reply:
It seems like there's still a per channel spin lock in use, has the change been pushed?